### PR TITLE
apache-arrow 0.4.0 (new formula)

### PR DESCRIPTION
**Pull request by hhatto**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

File: Formula/apache-arrow.rb [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    env = ENV["PATH"]
+    new_env = []
+    env.split(":").each do |value|
+      new_env.push value unless value.include? "shims/super"
+    end
+    new_env.push "/usr/local/bin"
+    ENV.store "PATH", new_env.join(":")
+    chdir "cpp"
+    build_type = "Release"
+    mkdir build_type.downcase
```
Comment (cosmo0920) [inline Formula/apache-arrow.rb]: We can use the following mkdir with do block notation:

```ruby
mkdir build_type.downcase do
  system "cmake", "-D", "CMAKE_BUILD_TYPE=#{build_type}", "-D", "CMAKE_INSTALL_PREFIX:PATH=#{prefix}", ".."
  # And building arrow
end
```

File: Formula/apache-arrow.rb [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  depends_on "boost"
+  depends_on "cmake" => :build
+
+  def install
+    env = ENV["PATH"]
+    new_env = []
+    env.split(":").each do |value|
+      new_env.push value unless value.include? "shims/super"
+    end
+    new_env.push "/usr/local/bin"
```
Comment (cosmo0920) [inline Formula/apache-arrow.rb]: Any chance to use `ENV.prepend_path`?

**Comment by hhatto** [comment]
Thank you for code review. How about this?

File: Formula/apache-arrow.rb [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  desc "Apache Arrow"
+  homepage "https://arrow.apache.org/"
+  url "https://github.com/apache/arrow/archive/apache-arrow-0.4.0.tar.gz"
+  sha256 "0ce9f47e42735d8d48dd83f9ff6a1612c5f2a8e846f3ae23595697e3519dd9b8"
+
+  head "https://github.com/apache/arrow.git"
+
+  depends_on "boost"
+  depends_on "cmake" => :build
+
```
Comment (cosmo0920) [inline Formula/apache-arrow.rb]: Apache arrow need to c++11 compiler option?

It might be needed to add `needs :cxx11` .

And we can use `Env.cxx11` in `#install`

File: Formula/apache-arrow.rb [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+  def install
+    ENV.prepend_path "PATH", "/usr/local/bin"
+    chdir "cpp"
+    build_type = "Release"
+    mkdir build_type.downcase do
+      system "cmake", "-D", "CMAKE_BUILD_TYPE=#{build_type}", "-D", "CMAKE_INSTALL_PREFIX:PATH=#{prefix}", ".."
```
Comment (cosmo0920) [inline Formula/apache-arrow.rb]: I guess that `*std_cmake_args` contains `CMAKE_INSTALL_PREFIX`.

File: Formula/apache-arrow.rb (hunk 3 above)
Comment (hhatto) [inline Formula/apache-arrow.rb]: > Apache arrow need to c++11 compiler option?

yes

> It might be needed to add needs :cxx11 .
And we can use Env.cxx11 in #install

OK. Thanks 👍

File: Formula/apache-arrow.rb (hunk 4 above)
Comment (hhatto) [inline Formula/apache-arrow.rb]: Thanks.

contains `CMAKE_BUILD_TYPE` too.
http://www.rubydoc.info/github/Homebrew/brew/Formula%3Astd_cmake_args

**Comment by hhatto** [comment]
test this please

File: Formula/apache-arrow.rb [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+  head "https://github.com/apache/arrow.git"
+
+  depends_on "boost"
+  depends_on "cmake" => :build
+
+  needs :cxx11
+
+  def install
+    ENV.prepend_path "PATH", "/usr/local/bin"
```
Comment (dunn) [inline Formula/apache-arrow.rb]: Why is this required?

File: Formula/apache-arrow.rb [hunk 6]
```diff
@@ -0,0 +1,36 @@
+class ApacheArrow < Formula
+  desc "Apache Arrow"
```
Comment (dunn) [inline Formula/apache-arrow.rb]: A more informative description would be nice.

Comment (cosmo0920) [inline Formula/apache-arrow.rb]: How about use "Powering Columnar In-Memory Analytics"?

File: Formula/apache-arrow.rb (hunk 5 above)
Comment (hhatto) [inline Formula/apache-arrow.rb]: [CMake's `find_program(ccache, )` ](https://github.com/apache/arrow/blob/master/cpp/CMakeLists.txt#L58) search for `ccache` regardless of `ENV["PATH"]`.
In order to avoid this case, added a `/usr/local/bin` path. I do not know if this is the best way.

Comment (dunn) [inline Formula/apache-arrow.rb]: Does the formula depend on ccache? If so, it should be added as a dependency.  Then Homebrew will make sure it's in the PATH.

**Comment by wesm** [comment]
Can you update to 0.4.1 (tagged last week)?

**Comment by hhatto** [comment]
> Can you update to 0.4.1 (tagged last week)?

I open a new PR.

File: Formula/apache-arrow.rb (hunk 5 above)
Comment (hhatto) [inline Formula/apache-arrow.rb]: Strictly speaking, ccache is not required. Because [this problem](https://github.com/apache/arrow/pull/751) occurs, I think that use `depends_on "ccache" => :recommended` now.
What do you think about this idea?

Comment (dunn) [inline Formula/apache-arrow.rb]: What happens when it's not built with ccache and that `PATH` line is not used?

Comment (hhatto) [inline Formula/apache-arrow.rb]: > What happens when it's not built with ccache and that PATH line is not used?

In that case build it will succeed, because not need for ccache's PATH.

Comment (dunn) [inline Formula/apache-arrow.rb]: If it's not necessary then, let's remove both, thanks!
